aosp_diff/preliminary/hardware/interfaces/09_0009-Add-the-memtrack-AIDL-HAL-implement.patch

```diff
@@ -0,0 +1,206 @@
+ memtrack/aidl/default/Memtrack.cpp        | 73 ++++++++++++++++++++++-
+ memtrack/aidl/default/Memtrack.h          | 15 +++++
+ memtrack/aidl/default/main.cpp            |  5 +-
+ memtrack/aidl/default/memtrack-default.rc |  2 +-
+ 5 files changed, 94 insertions(+), 4 deletions(-)
+
+diff --git a/memtrack/aidl/default/Android.bp b/memtrack/aidl/default/Android.bp
+index 7a7feea5bfa3..5fd86f1c3881 100644
+--- a/memtrack/aidl/default/Android.bp
++++ b/memtrack/aidl/default/Android.bp
+@@ -30,6 +30,9 @@ cc_binary {
+     shared_libs: [
+         "libbase",
+         "libbinder_ndk",
++        "liblog",
++        "libhardware",
++        "libutils",
+         "android.hardware.memtrack-V1-ndk_platform",
+     ],
+     srcs: [
+diff --git a/memtrack/aidl/default/Memtrack.cpp b/memtrack/aidl/default/Memtrack.cpp
+index 49a6582989ca..1c4c3fcf94bb 100644
+--- a/memtrack/aidl/default/Memtrack.cpp
++++ b/memtrack/aidl/default/Memtrack.cpp
+@@ -14,12 +14,28 @@
+  * limitations under the License.
+  */
+ 
++#define LOG_TAG "memtrack"
++
+ #include "Memtrack.h"
+ 
++#include <log/log.h>
++
++#include <hardware/hardware.h>
++#include <hardware/memtrack.h>
++
+ namespace aidl {
+ namespace android {
+ namespace hardware {
+ namespace memtrack {
++namespace impl {
++namespace example {
++
++Memtrack::Memtrack(const memtrack_module_t *module) : mModule(module) {
++    if (mModule)
++        mModule->init(mModule);
++}
++
++Memtrack::~Memtrack() {}
+ 
+ ndk::ScopedAStatus Memtrack::getMemory(int pid, MemtrackType type,
+                                        std::vector<MemtrackRecord>* _aidl_return) {
+@@ -31,16 +47,71 @@ ndk::ScopedAStatus Memtrack::getMemory(int pid, MemtrackType type,
+         return ndk::ScopedAStatus(AStatus_fromExceptionCode(EX_UNSUPPORTED_OPERATION));
+     }
+     _aidl_return->clear();
++
++    size_t temp = 0;
++    size_t *size = &temp;
++    int ret = 0;
++
++    if (mModule->getMemory == nullptr)
++    {
++	return ndk::ScopedAStatus::ok();
++    }
++    ret = mModule->getMemory(mModule, pid, static_cast<memtrack_type>(type),
++            NULL, size);
++    if (ret == 0)
++    {
++        memtrack_record *legacy_records = new memtrack_record[*size];
++        ret = mModule->getMemory(mModule, pid,
++                static_cast<memtrack_type>(type), legacy_records, size);
++        if (ret == 0)
++        {
++            for(size_t i = 0; i < *size; i++)
++            {
++		MemtrackRecord record = {
++			.flags = static_cast<int32_t>(legacy_records[i].flags),
++			.sizeInBytes = static_cast<int64_t>(legacy_records[i].size_in_bytes),
++		};
++		_aidl_return->emplace_back(record);
++            }
++        }
++        delete[] legacy_records;
++    }
++
+     return ndk::ScopedAStatus::ok();
+ }
+ 
+ ndk::ScopedAStatus Memtrack::getGpuDeviceInfo(std::vector<DeviceInfo>* _aidl_return) {
+     _aidl_return->clear();
+-    DeviceInfo dev_info = {.id = 0, .name = "virtio_gpu"};
++    DeviceInfo dev_info = {.id = 0, .name = "i915"};
```

Comment:
does it mean only mentrack only exist for CiV GVT-d and SRIOV?

Reply:
In fact this line: DeviceInfo dev_info = {.id = 0, .name = "i915"}; is needed for pass the VTS test.
Before someone add the virtio_gpu name to pass the VTS test.
I change it to i915 as we implement the AIDL hal and kernel have the memtrack function patch.

Reply:
i have checked on virtio, there is no kernel related memtrack function enabled.
Although the hal service is working, but there is no memtrack EGL info:
caas:/ # dumpsys meminfo |grep EGL                                             
          0K: EGL mtrack
          0K: EGL mtrack
caas:/ # service list |grep mem
10	android.hardware.memtrack.IMemtrack/default: [android.hardware.memtrack.IMemtrack]
119	meminfo: []
120	memtrack.proxy: [android.hardware.memtrack.IMemtrack]
caas:/ #